src/main/java/com/google/devtools/build/lib/actions/Spawns.java

```diff
@@ -152,21 +155,72 @@ public static Duration getTimeout(Spawn spawn) throws ExecException {
    * defaultTimeout, or {@code Duration.ZERO} if that is null.
    */
   public static Duration getTimeout(Spawn spawn, Duration defaultTimeout) throws ExecException {
-    String timeoutStr = spawn.getExecutionInfo().get(ExecutionRequirements.TIMEOUT);
-    if (timeoutStr == null) {
-      return defaultTimeout == null ? Duration.ZERO : defaultTimeout;
+    Duration timeout = null;
+
+    for (Map.Entry<String, String> entry: spawn.getExecutionInfo().entrySet()) {
+        String value = getTimeoutValueFromExecutionInfoKeyOrValue(spawn,entry);
+        if (value != null) {
+            if (timeout != null) {
+                throw new UserExecException(
+                    FailureDetail.newBuilder()
+                        .setMessage("multiple timeout values specified")
+                        .setSpawn(FailureDetails.Spawn.newBuilder().setCode(Code.DUPLICATE_TIMEOUT_TAGS))
+                        .build());
+            }
+            timeout = parseTimeoutValue(value);
+        }
+    }
+
+    if (timeout == null) {
+        timeout = defaultTimeout == null ? Duration.ZERO : defaultTimeout;
     }
+    return timeout;
+}
+
+/**
+ * Retrieves the timeout value from the execution info key or value.
+ * Timeout may either be specified as a tag on format timeout:<int>
+ * or as a key-value pair when specified as an execution_requirement.
+ */
+private static String getTimeoutValueFromExecutionInfoKeyOrValue(Spawn spawn, Map.Entry<String, String> entry) throws ExecException {
+    ParseableRequirement requirement = ExecutionRequirements.TIMEOUT;
+    String value = null;
+
+    if (entry.getKey().equals("timeout")) {
+        value = entry.getValue();
+    } else {
+        try {
+            value = requirement.parseIfMatches(entry.getKey());
+        } catch (ValidationException e) {
+            throw new UserExecException(
+                e,
+                FailureDetail.newBuilder()
+                    .setMessage(
+                        String.format(
+                            "%s has a '%s' tag, but its value '%s' didn't pass validation: %s",
+                            spawn.getTargetLabel(), requirement.userFriendlyName(), e.getTagValue(), e.getMessage()))
+                    .setSpawn(FailureDetails.Spawn.newBuilder().setCode(Code.INVALID_TIMEOUT_TAG))
```

Comment:
I think it's fine to reuse `Code.TIMEOUT` here and let the error message carry the additional information.